kong/plugins/rate-limiting/handler.lua

```diff
@@ -122,84 +122,131 @@ function RateLimitingHandler:access(conf)
-          window = current_window
-          remaining = current_remaining
-
-          if not timestamps then
-            timestamps = timestamp.get_timestamps(current_timestamp)
+          if not limit or (current_remaining < remaining)
+                      or (current_remaining == remaining and
+                          current_window > window)
+          then
+            limit = current_limit
+            window = current_window
+            remaining = current_remaining
+
+            if not timestamps then
+              timestamps = timestamp.get_timestamps(current_timestamp)
+            end
+
+            reset = max(1, window - floor((current_timestamp - timestamps[k]) / 1000))
           end
 
-          reset = max(1, window - floor((current_timestamp - timestamps[k]) / 1000))
+          headers[X_RATELIMIT_LIMIT[k]] = current_limit
+          headers[X_RATELIMIT_REMAINING[k]] = current_remaining
         end
 
-        headers[X_RATELIMIT_LIMIT[k]] = current_limit
-        headers[X_RATELIMIT_REMAINING[k]] = current_remaining
+        headers[RATELIMIT_LIMIT] = limit
+        headers[RATELIMIT_REMAINING] = remaining
+        headers[RATELIMIT_RESET] = reset
+      end
+
+      -- If limit is exceeded, terminate the request
+      if stop then
+        headers = headers or {}
+        headers[RETRY_AFTER] = reset
+        return kong.response.error(429, "API rate limit exceeded", headers)
       end
 
-      headers[RATELIMIT_LIMIT] = limit
-      headers[RATELIMIT_REMAINING] = remaining
-      headers[RATELIMIT_RESET] = reset
+      if headers then
+        kong.response.set_headers(headers)
+      end
     end
 
-    -- If limit is exceeded, terminate the request
-    if stop then
-      headers = headers or {}
-      headers[RETRY_AFTER] = reset
-      return kong.response.error(429, "API rate limit exceeded", headers)
+    local ok, err = timer_at(0, increment, conf, limits, identifier, current_timestamp, 1)
+    if not ok then
+      kong.log.err("failed to create timer: ", err)
     end
 
-    if headers then
-      kong.response.set_headers(headers)
+  elseif window_type == "sliding" then
+    local limit = conf.limit
+    local window_size = conf.window_size
+
+    if conf.policy ~= 'redis' then
+      return error("invalid policy for sliding window")
+    end
+
+    local current_usage, reset, err = policies[conf.policy].sliding_window_usage(conf, identifier, window_size)
+    if err then
+      if not fault_tolerant then
+        return error(err)
+      end
+
+      kong.log.err("failed to get sliding_window_usage: ", tostring(err))
+    end
+
+    if current_usage and reset then
+      local remaining = max(0, limit - current_usage)
+      local stop = current_usage > limit
+      local headers
+      if not conf.hide_client_headers then
+        headers = {}
+
+        headers[RATELIMIT_LIMIT] = limit
+        headers[RATELIMIT_REMAINING] = remaining
+        headers[RATELIMIT_RESET] = reset
+      end
```

Comment:
I think this is still a fixed window of time. The window is the TTL on a key, which never moves. So when the TTL expires after the fixed amount of time, the counter resets.

Reply:
IMHO TTL is fixed but the window of time is not. 
For example, if the time unit is 1 minute, fixed window counter guarantees an average rate in window [00:00, 00:01), [00:01,00:02), etc.
Suppose the limit is 10 per minute. 
In fixed window if 10 requests arives at 00:00:59 and more 10 requests arives at 00:01:01 we have 20 alowed requests in the interval of 3 seconds.   
With redis TTL it is not possible. At the same cenario only the first 10 requests are allowed. The last 10 recives 429 status code. The window limit slides between time window [00:00, 00:01) and [00:01,00:02) 
But I'm open to suggestions

Reply:
I think you would have to keep timestamps of each request in a sorted set.
and on each request drop the timestamps between 0 and (now - window_time), add now. the count of items in the set is how many requests have been made in the window.
if the number is greater than the configured value, the limit has been reached

This would make the window sliding over the points in time as the requests are made.